### PR TITLE
Export cpg as json via Neo4j OGM

### PR DESCRIPTION
**Pull request by peckto**
This PR adds the functionality to export the cpg as json.
The serialization is done via the Neo4j OGM.
Basically the same as when importing to Neo4j Database, just without Database.
So the json export will suffer from known problems, like recursion and bigInt.

Json format might be useful for machine learning or other data science projects.

Proposed json format:
```json
{
"nodes": [],
"edges": []
}
```
Example json file: 
[cpg.txt](https://github.com/Fraunhofer-AISEC/cpg/files/11757931/cpg.txt)

- [x] compare json export with data from Neo4j Database import
- [x] add test case
- [x] support pre-save hooks

### CODE REVIEW

**Comment by peckto** [comment]
The current implementation does not create any Neo4j Session object, which allows the json export to be used without a connection to the database server, which is of course desirable. 
This has unfortunately the side effect, that pre-save hooks are not executed, because they are configured in the session.
I have currently two ideas how this could be fixed:
a) create some dummy session, without actual database connection
b) configure pre-save hooks without session

**Comment by KuechA** [comment]
Related issues: #247 #776  #935 . We may want to enable customizing the output format somehow.

**Comment by peckto** [comment]
> We may want to enable customizing the output format somehow

Good point! I created a separate function to export in the proposed json format (`exportSimpleJson`) .
In future, here one could step in to support other formats.

**Comment by konradweiss** [comment]
- [ ] Make a pass `PrepareForNeo4jSerializationPass`
- [ ] Turn it on as a default pass for the Neo4j module.

File: cpg-neo4j/README.md [hunk 1]
```diff
@@ -14,57 +14,95 @@ Build (and install) a distribution using Gradle
...
 ## Known issues:
 
 - While importing sufficiently large projects with the parameter <code>--save-depth=-1</code> 
         a <code>java.lang.StackOverflowError</code> may occur.
     - This error could be solved by increasing the stack size with the JavaVM option: <code>-Xss4m</code>
+    - Otherwise the depth must be limited (e.g. 3 or 5)
 
 - While pushing a constant value larger than 2^63 - 1 a <code>java.lang.IllegalArgumentException</code> occurs.
-
-## Experimental Languages
```
Comment (konradweiss) [inline cpg-neo4j/README.md:58]: Why did we remove the lines for experimental languages?

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt [hunk 2]
```diff
@@ -232,6 +253,94 @@ class Application : Callable<Int> {
     private val passList: List<String>
         get() = passClassList.mapNotNull { it.simpleName }
 
+    private val packages: Array<String> =
+        arrayOf("de.fraunhofer.aisec.cpg.graph", "de.fraunhofer.aisec.cpg.frontends")
+
+    fun serializeCpgViaOGM(
```
Comment (konradweiss) [inline cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt:259]: I would document this function for better maintainability. Also I would rename some function as I was confused and that might happen to others IMO. Here `translateCPGToOGMBuilders`?

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt [hunk 3]
```diff
@@ -232,6 +253,94 @@ class Application : Callable<Int> {
...
+        val newNodeBuilders =
+            (getNewNodeBuilders[compiler] as? ArrayList<*>)?.filterIsInstance<DefaultNodeBuilder>()
+        val newRelationshipBuilders =
+            (getNewRelationshipBuilders[compiler] as? ArrayList<*>)?.filterIsInstance<
+                DefaultRelationshipBuilder
+            >()
+        return newNodeBuilders to newRelationshipBuilders
+    }
+
+    fun exportSimpleJson(
```
Comment (konradweiss) [inline cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt:290]: Here `buildJsonGraph`?

**Review by konradweiss — CHANGES_REQUESTED** [review]
I would change some names and add more explicit explanations in the form of function comments. Just to make it a bit more maintainable. I did also think whether or not it would be good to extract the export to json functionality into its own class to not overload the `Application.kt` but for now its fine so that is optional.

File: cpg-neo4j/README.md (hunk 1 above)
Comment (peckto) [reply cpg-neo4j/README.md:58]: The option does no longer exists. As far as I understand, experimental language frontends are treated as any other frontend, as all are now optional?

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt (hunk 2 above)
Comment (peckto) [reply cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt]: done

File: cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt (hunk 3 above)
Comment (peckto) [reply cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt]: done

File: cpg-neo4j/README.md (hunk 1 above)
Comment (KuechA) [reply cpg-neo4j/README.md:58]: Correct. Each frontend is basically optional

**Review by konradweiss — APPROVED** [review]

**Pull request merged**